### PR TITLE
Refine PM2.5 DA for the RRFS_SD model 

### LINKED ISSUE #606: Refine PM2.5 DA for RRFS-SD
Refine PM2.5 DA for the RRFS-SD model by using surface type and adding a few new namelist variables.

### PR DESCRIPTION
**Pull request by hongli-wang**
**Description**

Refine the PM2.5 DA for the RRFS_SD model by use of veg_type, which is used to decide whether the obs is in urban area or not. Different thresholds for innovations outside/inside urban areas will be used.  Add new namelist parameters, such as threshold for innovations, anowbufr type 
Read in station terrain height, PM10 et al if extended BUFR format for anow air quality data is used 

Fixes #606 

**Type of change**

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

  
**Checklist**

- [X ] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

**DUE DATE for this PR is 9/22/2023.** If this PR is not merged into develop by this date, the PR will be closed and returned to the developer.

### CODE REVIEW

**Comment by hu5970** [comment]
@hongli-wang Could you suggest two reviewers for this PR?

**Comment by hongli-wang** [comment]
> @hongli-wang Could you suggest two reviewers for this PR?

@hu5970 Previously, Ting and Cory reviewed the initial development of PM2.5 DA for RRFS_SD. They are the best to review this update.

**Comment by RussTreadon-NOAA** [comment]
Cory is not available as a peer reviewer.  Cory is a member of the Handling Review team.  Handling Review team members process PRs.  We need two peer reviewers not from the Handling Review team.

**Comment by hongli-wang** [comment]
> Cory is not available as a peer reviewer. Cory is a member of the Handling Review team. Handling Review team members process PRs. We need two peer reviewers not from the Handling Review team.

@RussTreadon-NOAA Thanks for letting me know. Guoqing @guoqing-noaa can help to review the code. Thanks.

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang can you start by running the GSI regression tests on Hera? I will also run them on WCOSS2 once we know the all pass on Hera. Thanks!

**Comment by hongli-wang** [comment]
> > Cory is not available as a peer reviewer. Cory is a member of the Handling Review team. Handling Review team members process PRs. We need two peer reviewers not from the Handling Review team.
> 
> @RussTreadon-NOAA Thanks for letting me know. Guoqing @guoqing-noaa can help to review the code. Thanks.

**Comment by hongli-wang** [comment]
> > Cory is not available as a peer reviewer. Cory is a member of the Handling Review team. Handling Review team members process PRs. We need two peer reviewers not from the Handling Review team.
> 
> @RussTreadon-NOAA Thanks for letting me know. Guoqing @guoqing-noaa can help to review the code. Thanks.

Youhua (youhua.tang@noaa.gov) can also review the code. Thanks.

File: src/gsi/chemmod.f90
Comment (TingLei-NOAA) [inline src/gsi/chemmod.f90]: Would add an explanation for =.false.?

File: src/gsi/chemmod.f90
Comment (TingLei-NOAA) [inline src/gsi/chemmod.f90]: documentation  for this variable

File: src/gsi/read_anowbufr.f90
Comment (TingLei-NOAA) [inline src/gsi/read_anowbufr.f90]: why don't directly use r180 (if r180 is not defined,  it could be defined along with r360).  The extra r180 will avoid the multiplication for 0.5*r360 for each obs.

**Comment by hongli-wang** [comment]
> @hongli-wang can you start by running the GSI regression tests on Hera? I will also run them on WCOSS2 once we know the all pass on Hera. Thanks!

 A bit busy this week. I will do it as soon as I can.

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90]: > Would add an explanation for =.false.?

Thanks for the comments. I will revise soon.

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang no big rush, we have a few weeks.

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [inline src/gsi/chemmod.f90:111]: This comment is not related to the code. But just curious whey bumping the max value from 150 to 3000?

File: src/gsi/read_anowbufr.f90
Comment (guoqing-noaa) [inline src/gsi/read_anowbufr.f90:166]: Suggest reverse the if and else block, i.e:  if anowbufr_ext then; else ...  (although they function the same, the latter reads more straightforward :) )

File: src/gsi/read_anowbufr.f90
Comment (guoqing-noaa) [inline src/gsi/read_anowbufr.f90:165]: Add some words to output that this is the anowbufr_ext type?

File: src/gsi/read_anowbufr.f90
Comment (guoqing-noaa) [inline src/gsi/read_anowbufr.f90:187]: same comments as the above

File: src/gsi/read_anowbufr.f90
Comment (guoqing-noaa) [inline src/gsi/read_anowbufr.f90]: remove two spaces to align the indentation

File: src/gsi/read_anowbufr.f90
Comment (guoqing-noaa) [inline src/gsi/read_anowbufr.f90]: For .not. anowbufr_ext, only read the header and don't need to read observations?

File: src/gsi/setuppm2_5.f90
Comment (guoqing-noaa) [inline src/gsi/setuppm2_5.f90]: remove line 782 if not needed

File: src/gsi/setuppm2_5.f90
Comment (guoqing-noaa) [inline src/gsi/setuppm2_5.f90]: Is the change from 1.0 to 2.0 a bug fix or a threshold change? For the latter, will it be better to use a parameter to control this?

File: src/gsi/setuppm2_5.f90
Comment (guoqing-noaa) [inline src/gsi/setuppm2_5.f90]: Do you want to remove line 796?

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [inline src/gsi/chemmod.f90:97]: I searched this PR, it looks like we don't need separate berror_fv3_sd_regional and berror_fv3_cmaq_regional?

berror_fv3_sd_regional is only used at Line 460 of src/gsi/m_berror_stats_reg.f90 and there is no difference if we use berror_fv3_cmaq_regional instead. Is it possible to just use berror_fv3_cmaq_regional, or combine them to one logical variable, such as berror_fv3_aero_regional?

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [inline src/gsi/chemmod.f90:168]: "coarsepm" is added here. However it is never used. And we have to modify Line 282 of src/gsi/setuppm2_5.f90
 to remove the "coarsepm" contribution

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [inline src/gsi/chemmod.f90]: pm10_da_scheme is not used anywhere in this PR.

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90:111]: > This comment is not related to the code. But just curious whey bumping the max value from 150 to 3000?

@guoqing-noaa The default value may not consider the heavy fire event. The value of 3000 is a reasonable PM10 value when a heavy fire event or dust storm takes place.

**Comment by hongli-wang** [comment]
@TingLei-NOAA @guoqing-noaa 
Hi Ting and Guoqing, 
Thanks for your comments. I have taken almost all of your suggestions and modified the codes. Please review the code again. Thanks. Hongli

**Comment by TingLei-NOAA** [comment]
@hongli-wang   just maybe last two question: did you run your test for this PR with GSI at debug mode? :).  and did you finish the regression tests?

**Comment by hongli-wang** [comment]
> @hongli-wang just maybe last two question: did you run your test for this PR with GSI at debug mode? :). and did you finish the regression tests?

@TingLei-NOAA yes tested my code on debug mode. Not finish regression test yet, and I will do it.

**Comment by TingLei-NOAA** [comment]
@hongli-wang  Thanks.  Then, the success of your regression tests are the only thing I need to know to give my approval for this PR.   Thanks a lot for your efforts for keeping improving the codes.

**Comment by hongli-wang** [comment]
> @hongli-wang Thanks. Then, the success of your regression tests are the only thing I need to know to give my approval for this PR. Thanks a lot for your efforts for keeping improving the codes.

**Comment by TingLei-NOAA** [comment]
Hongli,
Will this help ?
https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests).
Regards,
Ting

______________________________

Ting Lei

Physical Scientist, Contractor with Lynker in support of

EMC/NCEP/NWS/NOAA

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

***@***.***

301-683-3624

**Comment by hongli-wang** [comment]
> Hongli, Will this help ? https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests). Regards, Ting
> […](#)

Thanks for the reply. 
I think this was because I was tryin on Jet. I will try on Hera when my master key is activated.

**Comment by hongli-wang** [comment]
> > Hongli, Will this help ? https://github.com/NOAA-EMC/GSI/wiki/GSI-Ctests-(regression-tests). Regards, Ting
> > […](#)
> 
> Thanks for the reply. I think this was because I was tryin on Jet. I will try on Hera when my master key is activated.

@TingLei-NOAA @CoryMartin-NOAA 
I submitted regression test on Hera yesterday. But don't get any output. I resubmit the jobs again. Here is my test dir: 
/scratch1/BMC/wrfruc/hwang/ctest/GSI/build

@CoryMartin-NOAA Did you run the regression test?  
Thanks, 
Hongli

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang not yet but I plan to early next week

**Comment by hongli-wang** [comment]
> @hongli-wang not yet but I plan to early next week

@CoryMartin-NOAA @TingLei-NOAA 
My regression tests are always pending on regression_wait. It seems no resource to perform the tests. Any suggestions for the tests on Hera? 
Thanks, 
Hongli

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang you may need to change the account you submit from? I'm not sure. I'm in a code sprint this week but I can try taking a look on Monday.

**Comment by TingLei-NOAA** [comment]
> > @hongli-wang not yet but I plan to early next week
> 
> @CoryMartin-NOAA @TingLei-NOAA My regression tests are always pending on regression_wait. It seems no resource to perform the tests. Any suggestions for the tests on Hera? Thanks, Hongli

@hongli-wang Could you point me to your runs on hera and I could take a look ?

**Comment by hongli-wang** [comment]
> > > @hongli-wang not yet but I plan to early next week
> > 
> > 
> > @CoryMartin-NOAA @TingLei-NOAA My regression tests are always pending on regression_wait. It seems no resource to perform the tests. Any suggestions for the tests on Hera? Thanks, Hongli
> 
> @hongli-wang Could you point me to your runs on hera and I could take a look ?

@TingLei-NOAA 
/scratch1/BMC/wrfruc/hwang/ctest/GSI/build
Thanks for looking into this.

**Comment by TingLei-NOAA** [comment]
@hongli-wang  first, seems a syntactic error in line 124 of /scratch1/BMC/wrfruc/hwang/ctest/GSI/regression/regression_var.sh .   And when you re-submit ctest, 
would you please use "-VV"  option to let ctest output more information for debug?

**Comment by hongli-wang** [comment]
> @hongli-wang first, seems a syntactic error in line 124 of /scratch1/BMC/wrfruc/hwang/ctest/GSI/regression/regression_var.sh . And when you re-submit ctest,
> would you please use "-VV" option to let ctest output more information for debug?

@TingLei-NOAA 
Thanks for catching the typo. The first job is running in queue now.

**Comment by hongli-wang** [comment]
> > @hongli-wang first, seems a syntactic error in line 124 of /scratch1/BMC/wrfruc/hwang/ctest/GSI/regression/regression_var.sh . And when you re-submit ctest,
> > would you please use "-VV" option to let ctest output more information for debug?
> 
> @TingLei-NOAA Thanks for catching the typo. The first job is running in queue now.

@TingLei-NOAA Most of regression tests failed. Please feel free to have a look: 
 /scratch1/BMC/wrfruc/hwang/ctest/GSI/build/log.1

**Comment by TingLei-NOAA** [comment]
@hongli-wang  Could let us know what the situations were for those failed tests?   For example, failure for scalability tests are not regarded as an essential problem. They could be ignored.

**Comment by CoryMartin-NOAA** [comment]
Running the regression tests on Cactus resulted in 5/9 tests failing.

```
The following tests FAILED:
	  3 - global_4denvar (Failed)
	  4 - hwrf_nmm_d2 (Failed)
	  5 - hwrf_nmm_d3 (Failed)
	  6 - rtma (Failed)
	  8 - netcdf_fv3_regional (Failed)
```

For `global_4denvar`, the lowproc runtime exceeds the threshold by less than one second.

For `hwrf_nmm_d2`, the hiproc runtime exceeds the threshold by 15 seconds.

For `hwrf_nmm_d3`, the lowproc runtime exceeds the threshold by 38 seconds.

The `rtma` test fails the lowproc runtime by 102 seconds.

The `fv3_regional` test fails because it exceeds the allowable memory usage 222324KB > 201740KB.

I don't think any of these are 'showstoppers' but this many tests failing is a bit concerning to me. How many fail on Hera / Orion?

@RussTreadon-NOAA I see a message in the log `ncdump: command not found` is there a module we need to load before running the regression tests or is that a known bug in the GSI regression tests on WCOSS2?

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , thanks for catching the WCOSS2 `ncdump: command not found`.   I agree that a module load or something is missing.  I never noticed this but then I never looked.   The robustness of the GSI repo and ctests is beginning to show cracks after 1+ year of no one serving a the central code manager.  Things slip through the cracks.

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang can you update to the head of GSI develop? I think the regression test issues I'm seeing on both Hera and Orion are related to issues with changes from #571 not being included here

**Comment by hongli-wang** [comment]
> @hongli-wang can you update to the head of GSI develop? I think the regression test issues I'm seeing on both Hera and Orion are related to issues with changes from #571 not being included here

@CoryMartin-NOAA yes, just rebased my code, will have some tests and run a regression test.

**Comment by hongli-wang** [comment]
> > @hongli-wang can you update to the head of GSI develop? I think the regression test issues I'm seeing on both Hera and Orion are related to issues with changes from #571 not being included here
> 
> @CoryMartin-NOAA yes, just rebased my code, will have some tests and run a regression test.

@CoryMartin-NOAA @TingLei-NOAA 

Regression tests failed due to GSI can't find correct intel lib, for example: 
/scratch1/BMC/wrfruc/hwang/ctest/tmpreg_rrfs_3denvar_glbens/rrfs_3denvar_glbens_loproc_updat/gsi.x: error while loading shared libraries: libmkl_intel_lp64.so.2
Dir on Hera: 
/scratch1/BMC/wrfruc/hwang/ctest

Any suggestions?

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang that's because the modulefiles changes in develop recently. Make sure your `develop` build is also up to date now that you've pulled in the recent updates.

I'm running the tests on Orion and Cactus now.

**Comment by hongli-wang** [comment]
> @hongli-wang that's because the modulefiles changes in develop recently. Make sure your `develop` build is also up to date now that you've pulled in the recent updates.
> 
> I'm running the tests on Orion and Cactus now.

@CoryMartin-NOAA 
Both develop and my branch are up to date and recompiled. Is there anything else I can do to keep the build to to date? 
Thanks for performing regression tests on Orion and Cactus.

**Comment by CoryMartin-NOAA** [comment]
Rerunning on Cactus:
All tests pass except `hwrf_nmm_d3` which fails the scalability test and `rtma` which fails the runtime maximum for a couple of the cases.

On Orion:
```
The following tests FAILED:
	  2 - [=[global_4dvar]=] (Failed)
	  5 - [=[hwrf_nmm_d3]=] (Failed)
	  6 - [=[rtma]=] (Failed)
	  7 - [=[rrfs_3denvar_glbens]=] (Failed)
	  8 - [=[netcdf_fv3_regional]=] (Failed)
```
The 4DVar test failed the scalability test.
The HWRF test barely failed the scalability test.
The RTMA test also failed the scalability test.
The RRFS test exceeded the runtime for one test by 3 seconds.
Finally, the FV3 regional test also failed for scalability.

I think given these results, the tests "pass"

**Comment by hongli-wang** [comment]
> Rerunning on Cactus: All tests pass except `hwrf_nmm_d3` which fails the scalability test and `rtma` which fails the runtime maximum for a couple of the cases.
> 
> On Orion:
> 
> ```
> The following tests FAILED:
> 	  2 - [=[global_4dvar]=] (Failed)
> 	  5 - [=[hwrf_nmm_d3]=] (Failed)
> 	  6 - [=[rtma]=] (Failed)
> 	  7 - [=[rrfs_3denvar_glbens]=] (Failed)
> 	  8 - [=[netcdf_fv3_regional]=] (Failed)
> ```
> 
> The 4DVar test failed the scalability test. The HWRF test barely failed the scalability test. The RTMA test also failed the scalability test. The RRFS test exceeded the runtime for one test by 3 seconds. Finally, the FV3 regional test also failed for scalability.
> 
> I think given these results, the tests "pass"

@CoryMartin-NOAA 
That is a good news. Thanks! 
I have the issue that GSI can't find the lib to run on Hera. The same issue if I ran GSI on Jet new. 

Do you think we can move forward, ask Ting and Guoqing to review and approve this PR? 

Thank again! 
Hongli

**Comment by CoryMartin-NOAA** [comment]
@guoqing-noaa @TingLei-NOAA can you re-review this? We can double check the regression tests at the end but I think all we are seeing here is machine instability. Thanks @hongli-wang for your work on this so far!

**Comment by CoryMartin-NOAA** [comment]
GitHub says 76 files changed, I think this needs updated to the head of develop again before it can be accurately reviewed.

**Comment by hongli-wang** [comment]
> GitHub says 76 files changed, I think this needs updated to the head of develop again before it can be accurately reviewed.

@CoryMartin-NOAA 
I noticed this too. It was caused my last rebase (not sure why it happened). I will rebase/merge the code again.

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90:97]: The current ones make it more clear to know which B file will be used, and the previous retros runs can repeat without any changes of GSI namelist parameter. This is good to users. 
Your suggestion will make the changes be general and good in code development. 
So what do you think?

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90:168]: Yes, you are correct. This was done by -1 in the do loop. For example, 
do i=2,naero_smoke_fv3-1 ! remove contribution from coarsepm

File: src/gsi/read_anowbufr.f90
Comment (hongli-wang) [reply src/gsi/read_anowbufr.f90:166]: Accepted. Thanks.

File: src/gsi/read_anowbufr.f90
Comment (hongli-wang) [reply src/gsi/read_anowbufr.f90:187]: Accepted. Thanks.

**Review by TingLei-NOAA — APPROVED** [review]

**Comment by hongli-wang** [comment]
> > GitHub says 76 files changed, I think this needs updated to the head of develop again before it can be accurately reviewed.
> 
> @CoryMartin-NOAA I noticed this too. It was caused my last rebase (not sure why it happened). I will rebase/merge the code again.

@CoryMartin-NOAA 
Remerged code to EMC develop branch and delete a unused namelist variable.  Could you please help to run regression tests if it is required? I can try if the reg tests work on Hera tomorrow.  Thanks.

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [reply src/gsi/chemmod.f90:97]: Do you plan to use different BEC for cmaq and sd? Or will they be the same as they are right now?

File: src/gsi/chemmod.f90
Comment (guoqing-noaa) [reply src/gsi/chemmod.f90:168]: Right, if we will remove coarsepm in src/gsi/setuppm2_5.f90 (line 281), why would we add it here?

**Comment by hongli-wang** [comment]
Please see my reply in the previous email.

**Comment by hongli-wang** [comment]
Different B files are used. The one for CMAQ has 70 aerosol species whereas
rrfs-ad only has 3 tracers.

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang at this point, I'll re-run the regression tests on WCOSS once we have approvals from the reviewers.

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90:168]: @guoqing-noaa 
This is because the dust PM10/AOD DA is not included in this PR. But we do consider EnVar to update the coarsepm, possibly in a next PR with FED DA.

File: src/gsi/chemmod.f90
Comment (hongli-wang) [reply src/gsi/chemmod.f90:97]: @guoqing-noaa 
Different B files are used. The one for CMAQ has 70 aerosol species whereas
rrfs-ad only has 3 tracers. I think a separate parameter is good for users. I have no problem in combining then if that is the way you prefer. Thanks.

Comment (guoqing-noaa) [reply src/gsi/chemmod.f90:97]: Got it. Thanks for the explanation!

**Review by guoqing-noaa — APPROVED** [review]
LGTM. Thanks for the discussions and changes!

**Comment by hongli-wang** [comment]
> @hongli-wang at this point, I'll re-run the regression tests on WCOSS once we have approvals from the reviewers.

@CoryMartin-NOAA 
Ting and Guoqing has approved this PR. Could you please run the regression test again? 

Thanks, 
Hongli

**Comment by CoryMartin-NOAA** [comment]
@hongli-wang already submitted them this morning on Cactus. I'll post when they are finished.

**Comment by CoryMartin-NOAA** [comment]
Only 2 regression tests fail on Cactus, and they are due to scalability, which are nonfatal. I think this is good to go now.

**Review by CoryMartin-NOAA — APPROVED** [review]

**Comment by hongli-wang** [comment]
> Only 2 regression tests fail on Cactus, and they are due to scalability, which are nonfatal. I think this is good to go now.

@CoryMartin-NOAA 
Thank you!

**Pull request merged**
